bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TableUtils.java

```diff
@@ -94,10 +95,13 @@ public static Long hash(Object obj, Node parent) {
                 }
                 return result;
             } else {
-                return (long) obj.hashCode();
+                return (long) obj.toString().hashCode();
             }
         } else {
-            return (long) obj.hashCode();
+            if (obj instanceof BmpStringValue) {
+                return (long) (obj.toString().concat(" ")).hashCode();
```

Comment:
This will create different behaviour for subclasses of `StringValue` class. What about instead of `space` we add a unique suffix/prefix that we know. If we can change `StringValue` classes `hashCode` without breaking the code that is also fine. We may have to discuss this with the runtime team.

Reply:
An idea for salt is a unique one based on the string value it self. We don't have to keep the salt.
Eg:   
run length coding or skipping even chars.
| string | salt1 | salt2 |
| ----- | -----| -----|
| abcd |  `a1b1c1d1` | `ac` | 
|aabbcc | `a2b2c2` | `abc` |

Reply:
but with this salting method still we will get same hash for 0 and "0". In order get unique hash for both I think we need unique salt values and we have to keep the salt values.

Reply:
I am thinking of a hash function using the value and the TypeTag attached with the value. So we can get rid of `instanceof` check and use a generic approach. 
For example, the hash can be like `31 * hash(TypeChecher.getType(obj).getTag()) + obj.hashCode()`; for line no 98 and keep the original code for line no 100. WDYT?

Reply:
Also in line 59, lets check if the obj is a `SimpleValue` instead of `RefValue`, then the String values will go inside the outermost If block